### ISSUE TITLE
Build documentation from docstrings (e.g. constraints)

### ISSUE THREAD

**Issue by tarskul** [spine-tools/SpineOpt.jl#649]
To streamline the development of new features and the documentation thereof, it could be interesting to build the documentation from the docstrings instead of making separate documentation files. This approach could also help to make sure that the documentation unit test does not fail (issue [#626](https://github.com/spine-tools/SpineOpt.jl/issues/626)).

I've already started an experiment with the constraints to look for possibilities for implementing this approach. From that experiment a few implementation considerations became clear.

The first is that the constraints are written in Latex which julia docstrings do not like due to the many escaped characters. One way around this is to escape the characters correctly. However that might be confusing when you are used to writing Latex regularly. There is another way which allows to write Latex within a special docstring provided by the Documenter package. However, that would require to add Documenter as a dependency of the SpineOpt tool.

Regardless of the Latex characters, there are also two ways to build the documentation from the docstring. The first is quite straightforward. Documenter allows to call the docstring documentation directly in your documentation file. Although that should be rather straightforward, the result might not look too great. That is unless we deviate from the typical structure where the function and the corresponding parameters are highlighted. Another approach is to add special fields to the docstring and search for those special fields. In the case of the constraints that means that there will at least be two fields: one for the formulation and one for a short description thereof. I propose to mark the fields with the #region format as used in the widely used code editor vs code.

An example of how a docstring could look like is the following:

```julia
@doc raw"""
    add_constraint_nodal_balance!(m::Model)

Balance equation for nodes.

#region formulation
\begin{aligned}
& v_{node\_injection}(n,s,t) \\
& + \sum_{\substack{(conn,n',d_{in},s,t) \in connection\_flow\_indices: \\ d_{out} == :to\_node}}
 v_{connection\_flow}(conn,n',d_{in},s,t)\\
& - \sum_{\substack{(conn,n',d_{out},s,t) \in connection\_flow\_indices: \\ d_{out} == :from\_node}}
 v_{connection\_flow}(conn,n',d_{out},s,t)\\
 & + v_{node\_slack\_pos}(n,s,t) \\
 & - v_{node\_slack\_neg}(n,s,t) \\
& \{>=,==,<=\} \\
& 0 \\
& \forall (n,s,t) \in node\_stochastic\_time\_indices: \\
& p_{balance\_type}(n) != balance\_type\_none \\
& \nexists ng \in groups(n) : balance\_type\_group \\
\end{aligned}
#endregion formulation

#region description
### [Nodal balance](@id constraint_nodal_balance)
In **SpineOpt**, [node](@ref) is the place where an energy balance is enforced. As universal aggregators,
they are the glue that brings all components of the energy system together. An energy balance is created for each [node](@ref) for all `node_stochastic_time_indices`, unless the [balance\_type](@ref) parameter of the node takes the value [balance\_type\_none](@ref balance_type_list) or if the node in question is a member of a node group, for which the [balance\_type](@ref) is [balance\_type\_group](@ref balance_type_list). The parameter [nodal\_balance\_sense](@ref) defaults to equality, but can be changed to allow overproduction ([nodal\_balance\_sense](@ref) [`>=`](@ref constraint_sense_list)) or underproduction ([nodal\_balance\_sense](@ref) [`<=`](@ref constraint_sense_list)).

The constraint consists of the [node injections](@ref constraint_node_injection), the net [connection\_flow](@ref)s and [node slack variables](@ref Variables).
#endregion description
"""
function add_constraint_nodal_balance!(m::Model)
    [...]
end
```

Summary of design decisions:
1. Latex in the docstring
    a. If we want to directly write latex equations we have to add the Documenter dependency to SpineOpt and use its docstrings (@doc raw""" """)
    b. If we want to keep normal docstrings, we'll have to correctly escape the special latex characters
2. Approach
    a. Use the docstring directly
        i. should be rather straightforward to implement
        ii. the content of the docstring will either deviate from the standard approach or will look very artificial in the documentation
    b. Find specific fields in the docstring
        i. the implementation requires a more serious script
        ii. flexible way to create the documentation with multiple fields: #region formulation ; #region description (#region used because the widely used vs code can recognise that field too)

**Comment by tarskul** [comment]
TLDR: Do we add the dependency on documenter to avoid the need for escaping characters?

After improvements, cleaning, bugfixes and whateverelsetheremightbe I managed to fully implement the feature to add documentation from the docstrings. Just like the drag and drop feature (#618), the documentation from docstring feature is also fully optional.

The most practical way to use the function is to simply write a markdownfile. You can make titles and the small texts that make it into a story. But as soon as you come to a part where you can insert the docstring of a function you can instead insert instructions. The instructions consist of the function name and the specific fields from the docstring that you would like to add. Each on a separate line. The instructions need to be incapsulated with #region instruction and #endregion instruction. For example:

```markdown
# Constraints
## Balancing
#region instruction
add_constraint_nodal_balance!
description
formulation
#endregion instruction
```
with
```julia
@doc raw"""
    add_constraint_nodal_balance!()

Adds the nodal balance constraint

    #region description
    This is the nodal balance constraint:
    #endregion description

    #region formulation
    '''math
    v_{node\_injection} = 0
    '''
    #endregion formulation
"""
function add_constraint_nodal_balance()
end
```

Will result in the markdown file
```markdown
# Constraints
## Balancing
This is the nodal balance constraint:

'''math
v_{node\_injection} = 0
'''
```

Note that it is important that the regions are indented in the docstring of the formulation. Also note that you can only add one instruction per instruction region. Further note that there is automatically a white space between different regions.

This approach implies that we can simply keep our current markdown file for the constraint documentation and we can gradually move the documentation from the markdown file to the docstring as we are working on certain constraints. It should also just work for any docstring but I have not tested that.

There are some more ways that these functions can be used but let's not dwell on that here.

In principle the code is ready to merge but we still have to make the decision whether we would like to add the dependency on documenter or not (the current example is with the dependency on documenter because we can use regular latex and just copy paste from the constraints.md file but I've also implemented the version without documenter in another branch; the code is actually the same except that we need to escape the characters correctly in the docstring: \\\\ instead of \\).
